File: 2024/Python/Day2/solutionB.py
def is_valid_increase(line):
    if (len(line) == 1):
        return True
    if (line[0] - line[1] >= -3 and line[0] - line[1] < 0): 
        return is_valid_increase(line[1:])
    return False

def is_valid_decrease(line):
    if (len(line) == 1):
        return True
    if (line[0] - line[1] <= 3 and line[0] - line[1] > 0): 
        return is_valid_decrease(line[1:])
    return False

count = 0
with open("../../data/day2.txt") as file:
    for line in file:
        l = [int(num) for num in line.rstrip('\n').split()]
        if (is_valid_increase(l) or is_valid_decrease(l)):
            count += 1
            continue
        
        dampener = False
        for i in range(0, len(l)):
            newList = l[:i] + l[i+1:]
            if (is_valid_increase(newList) or is_valid_decrease(newList)):
                dampener = True
                break

        if (dampener == True):
            count += 1


print(f"Solution: {count}")

File: 2024/JS/Day2/solutionB.js
const fs = require("fs");

const isValidIncrease = (line) => {
  if (line.length === 1) return true;
  if (line[0] - line[1] >= -3 && line[0] - line[1] < 0)
    return isValidIncrease(line.slice(1));
  return false;
};

const isValidDecrease = (line) => {
  if (line.length === 1) return true;
  if (line[0] - line[1] <= 3 && line[0] - line[1] > 0)
    return isValidDecrease(line.slice(1));
  return false;
};

fs.readFile("../../data/day2.txt", (err, data) => {
  if (err) {
    console.error(err);
    return;
  }

  const lines = data.toString().split("\n").slice(0, -1);
  const res = lines.reduce((acc, l) => {
    const line = l.split(" ");
    if (isValidIncrease(line) || isValidDecrease(line)) {
      return acc + 1;
    }

    let dampener = false;
    line.forEach((_n, i) => {
      const newLine = line.slice(0, i).concat(line.slice(i + 1));
      if (isValidIncrease(newLine) || isValidDecrease(newLine)) {
        dampener = true;
        return;
      }
    });

    if (dampener) return acc + 1;

    return acc;
  }, 0);

  console.log(`Result: ${res}`);
});
